From a1b2c3d4e5f60718293a4b5c6d7e8f90123456789 Mon Sep 17 00:00:00 2001
From: Tanya Singh <tanya_singh@accton.com>
Date: Thu, 24 Jul 2026 15:00:00 +0800
Subject: [PATCH] mtk: hostapd: increase netlink socket receive buffer size

Increase the dynamic VLAN netlink socket receive buffer to 256KB to
prevent dropped messages during heavy RADIUS exchanges (e.g. OpenRoaming
with large certificate chains and vendor-specific attributes). Without
this, the netlink socket used by full_dynamic_vlan_init() can hit
"No buffer space available" on recvmsg and lose RADIUS-assigned VLAN
events.

Signed-off-by: Tanya Singh <tanya_singh@accton.com>
---
 src/ap/vlan_full.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/ap/vlan_full.c b/src/ap/vlan_full.c
index 1a2b3c4d5..6e7f8a9b0 100644
--- a/src/ap/vlan_full.c
+++ b/src/ap/vlan_full.c
@@ -772,6 +772,7 @@ full_dynamic_vlan_init(struct hostapd_data *hapd)
 {
 	struct sockaddr_nl local;
 	struct full_dynamic_vlan *priv;
+	int rxbuf = 262144;
 
 	priv = os_zalloc(sizeof(*priv));
 	if (priv == NULL)
@@ -790,6 +791,7 @@ full_dynamic_vlan_init(struct hostapd_data *hapd)
 		os_free(priv);
 		return NULL;
 	}
+	setsockopt(priv->s, SOL_SOCKET, SO_RCVBUF, &rxbuf, sizeof(rxbuf));
 
 	os_memset(&local, 0, sizeof(local));
 	local.nl_family = AF_NETLINK;
